File: src/android/OnCellStratis.java
package com.oncell;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaWebView;
import org.apache.cordova.PluginResult;
import org.apache.cordova.PluginResult.Status;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import org.json.JSONObject;
import org.json.JSONArray;
import org.json.JSONException;

import java.lang.Runnable;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.HashMap;

import android.util.Log;
import android.content.Context;
import android.os.Handler;

import com.stratisiot.stratissdk.StratisSDK;
import com.stratisiot.stratissdk.constants.ServerEnvironment;
import com.stratisiot.stratissdk.model.Configuration;
import com.stratisiot.stratissdk.error.StratisError;
import com.stratisiot.stratissdk.listener.DeviceActivationEvent;
import com.stratisiot.stratissdk.listener.StratisDeviceAccessListener;
import com.stratisiot.stratissdk.listener.StratisDeviceActivationListener;
import com.stratisiot.stratissdk.listener.StratisDeviceDiscoveryListener;
import com.stratisiot.stratissdk.model.lock.StratisLock;
import com.stratisiot.stratissdk.model.lock.BLELock;

import com.bugsnag.android.*;

public class OnCellStratis extends CordovaPlugin {
    private static final String TAG = "OnCellStratis";
    private StratisSDK stratisSDK;
    private Collection<StratisLock> accessibleLocks;
    private HashSet<StratisLock> discoveredLocks;

    public OnCellStratis() {
        super();
    }

    @Override
    public void initialize(CordovaInterface cordova, CordovaWebView webView) {
        Log.d(TAG, "OnCellStratis plugin initialize");
        super.initialize(cordova, webView);
    }
    
    private enum Commands {
        initSDK, getLocks, scanLocks, activateLock
    }
    
    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        Commands cmd = Commands.valueOf(action);
        switch (cmd) {
            case initSDK:
                String serverEnvironmentString = args.getString(0);
                String accessToken = args.getString(1);
                String propertyId = args.getString(2);
                this.initSDK(serverEnvironmentString, accessToken, propertyId, callbackContext);
                return true;
            case getLocks:
                this.getLocks(callbackContext);
                return true;
            case scanLocks:
                this.scanLocks(callbackContext);
//                this.scanLocksAfterDelay(callbackContext, 5); // Use for testing purposes
                return true;
            case activateLock:
                String lockId = args.getString(0);
                String appointmentId = args.getString(1);
                this.activateLock(lockId, appointmentId, callbackContext);
                return true;
            default:
                return false;
        }
    }

    /* Begin command functions */

    public void initSDK(String serverEnvironmentString, String accessToken, String propertyId, CallbackContext callbackContext) {
        Log.d(TAG, "initSDK");

        if (stratisSDK == null) { // Only initialize once
            Bugsnag.leaveBreadcrumb("Initializing StratisSDK serverEnvironmentString: " + serverEnvironmentString + ", accessToken: " + accessToken + ", propertyId: " + propertyId);
            
            if (isValidServerEnvironment(serverEnvironmentString) && accessToken != null && propertyId != null) {
                MainActivity mainActivity = (MainActivity) cordova.getActivity();
                Context appContext = mainActivity.getApplicationContext();
                Map<String, String> logMap = new HashMap<String, String>();
                logMap.put("app", "OnCell");

                Configuration configuration = new Configuration(
                        ServerEnvironment.valueOf(serverEnvironmentString),
                        accessToken,
                        true,
                        logMap,
                        propertyId
                );

                stratisSDK = new StratisSDK(appContext, configuration);

                // return successfully
                callbackSuccess(callbackContext, null);
            } else { // return with error
                callbackError(callbackContext, "Missing or invalid parameters when initializing Stratis SDK");
            }
        } else {
            // if SDK is already initialized, just refresh the access token and property ID
            Bugsnag.leaveBreadcrumb("Refreshing Stratis SDK accessToken: " + accessToken + ", propertyId: " + propertyId);
            stratisSDK.setAccessToken(accessToken);
            stratisSDK.setPropertyID(propertyId);
            callbackSuccess(callbackContext, null);
        }
    }


    public void getLocks(CallbackContext callbackContext) {
        Log.d(TAG, "getLocks");
        stratisSDK.setDeviceAccessListener(new OnCellStratisDeviceAccessListener(callbackContext));
        stratisSDK.fetchAccessibleDevices();
    }

    private class OnCellStratisDeviceAccessListener implements StratisDeviceAccessListener {

        private CallbackContext callbackContext;

        public OnCellStratisDeviceAccessListener(CallbackContext cc) {
            super();
            this.callbackContext = cc;
        }

        @Override
        public void stratisDeviceAccessRequestCompleted(@NotNull StratisSDK stratisSDK, @NotNull Collection<? extends StratisLock> collection, @Nullable StratisError stratisError) {
            Log.d(TAG, "stratisDeviceAccessRequestCompleted");
            if (stratisError == null) {
                accessibleLocks = (Collection<StratisLock>) collection;
                JSONArray locksJson = new JSONArray();
                if (!accessibleLocks.isEmpty()) {
                    locksJson = getLocksAsJson(accessibleLocks);
                    Log.d(TAG, "accessibleLocks: " + locksJson.toString());
                    Bugsnag.leaveBreadcrumb("accessibleLocks: " + locksJson.toString());
                } else {
                    bugsnagNotify("No accessible locks found");
                    Log.d(TAG, "No accessible locks found");
                }
                callbackSuccess(callbackContext, locksJson);
            } else {
                callbackError(callbackContext, stratisError.getDebugMessage());
            }
        }
    }

    public void scanLocksAfterDelay(CallbackContext callbackContext, Integer secondsDelay) {
        // Used only for testing purposes
        Log.d(TAG, "Delaying scanLocks...");
        Runnable runnable = new Runnable() {
            @Override
            public void run() {
                Log.d(TAG, "Calling scanLocks now");
                    scanLocks(callbackContext);
            }
        };
        Handler handler = new Handler();
        handler.postDelayed(runnable, secondsDelay*1000);
    }


    public void scanLocks(CallbackContext callbackContext) {
        Log.d(TAG, "scanLocks");
        if (accessibleLocks != null) {
            discoveredLocks = new HashSet<StratisLock>();
            stratisSDK.setDeviceDiscoveryListener(new OnCellStratisDeviceDiscoveryListener(callbackContext));
            stratisSDK.discoverActionableDevices(accessibleLocks);
        } else {
            callbackError(callbackContext, "Cannot scan locks before getting lock authorization");
        }
    }

    private class OnCellStratisDeviceDiscoveryListener implements StratisDeviceDiscoveryListener {

        private CallbackContext callbackContext;

        public OnCellStratisDeviceDiscoveryListener(CallbackContext cc) {
            super();
            this.callbackContext = cc;
        }

        @Override
        public void stratisDiscoveredDevices(@NotNull StratisSDK stratisSDK, @NotNull Collection<? extends StratisLock> collection) {
            Log.d(TAG, "stratisDiscoveredDevices");
            Log.d(TAG, collection.toString());
            discoveredLocks.addAll(collection);
        }

        @Override
        public void stratisDiscoveryCompleted(@NotNull StratisSDK stratisSDK) {
            Log.d(TAG, "stratisDiscoveryCompleted");
            JSONArray locksJson = new JSONArray();
            if (!discoveredLocks.isEmpty()) {
                ArrayList sortedDiscoveredLocks = new ArrayList<>(discoveredLocks);
                // Sort in descending order with highest RSSI (closest lock) first
                Comparator<BLELock> compareByRssi = (BLELock l1, BLELock l2) -> l2.getRssi().intValue() - l1.getRssi().intValue();
                Collections.sort(sortedDiscoveredLocks, compareByRssi);
                locksJson = getLocksAsJson(sortedDiscoveredLocks);
                Log.d(TAG, "discoveredLocks: " + locksJson.toString());
                Bugsnag.leaveBreadcrumb("discoveredLocks: " + locksJson.toString());
            } else {
                bugsnagNotify("No scanned locks discovered");
                Log.d(TAG, "No scanned locks discovered");
            }
            callbackSuccess(callbackContext, locksJson);
        }

        @Override
        public void stratisDiscoveryEncounteredError(@NotNull StratisSDK stratisSDK, @NotNull StratisError stratisError) {
            Log.e(TAG, "stratisDiscoveryEncounteredError");
            callbackError(callbackContext, stratisError.getDebugMessage());
        }
        
        @Override
        public void stratisDiscoveryDevicesOutOfRange(@NotNull StratisSDK stratisSDK, @NotNull Collection<? extends StratisLock> collection) {
            
        }

        @Override
        public void stratisDiscoveryUpdatedRSSI(@NotNull StratisSDK stratisSDK, @NotNull Collection<? extends StratisLock> collection) {
            Log.d(TAG, "stratisDiscoveryCompleted");
            JSONArray locksJson = new JSONArray();
            if (!discoveredLocks.isEmpty()) {
                ArrayList sortedDiscoveredLocks = new ArrayList<>(discoveredLocks);
                // Sort in descending order with highest RSSI (closest lock) first
                Comparator<BLELock> compareByRssi = (BLELock l1, BLELock l2) -> l2.getRssi().intValue() - l1.getRssi().intValue();
                Collections.sort(sortedDiscoveredLocks, compareByRssi);
                locksJson = getLocksAsJson(sortedDiscoveredLocks);
                Log.d(TAG, "discoveredLocks: " + locksJson.toString());
                Bugsnag.leaveBreadcrumb("discoveredLocks: " + locksJson.toString());
            } else {
                bugsnagNotify("No scanned locks discovered");
                Log.d(TAG, "No scanned locks discovered");
            }
            callbackSuccess(callbackContext, locksJson);
          }
    }


    public void activateLock(String lockId, String appointmentId, CallbackContext callbackContext) {
        Log.d(TAG, "activateLock");
        Bugsnag.leaveBreadcrumb("lockId: " + lockId);
        // appointmentId is no longer necessary for the Stratis SDK, but I'm leaving it here for debugging purposes
        Bugsnag.leaveBreadcrumb("appointmentId: " + appointmentId);

        if (discoveredLocks != null ) {
            StratisLock deviceToActivate = null;
            for (StratisLock lock : discoveredLocks) {
                if (lock.getIdentifier().equals(lockId)) {
                    deviceToActivate = lock;
                }
            }
            if (deviceToActivate != null) {
                OnCellStratisDeviceActivationListener listener = new OnCellStratisDeviceActivationListener(callbackContext);
                deviceToActivate.setListener(listener);
                deviceToActivate.activate();
                // return error to JS and send log to Bugsnag if we haven't unlocked device in 20 seconds
                listener.startActivationTimer(20);
            } else {
                callbackError(callbackContext, "Cannot find lock to activate");
            }
        } else {
            callbackError(callbackContext, "Cannot activate lock before scanning for locks");
        }
    }

    private class OnCellStratisDeviceActivationListener implements StratisDeviceActivationListener {

        private CallbackContext callbackContext;
        private Boolean responseSent = false;

        public OnCellStratisDeviceActivationListener(CallbackContext cc) {
            super();
            this.callbackContext = cc;
        }

        public void startActivationTimer(Integer maxSeconds) {
            // if we haven't sent a response to the front-end after X seconds, return error and log to Bugsnag
            Runnable runnable = new Runnable() {
                @Override
                public void run() {
                    Log.d(TAG, "activateLock responseSent: " + responseSent);
                    if (!responseSent) {
                        callbackError(callbackContext,"Lock activation took too long to complete");
                    }
                }
            };
            Handler handler = new Handler();
            handler.postDelayed(runnable, maxSeconds*1000);
        }

        @Override
        public void stratisDeviceActivationDidPostEvent(@NotNull StratisLock stratisLock, @NotNull DeviceActivationEvent deviceActivationEvent, @Nullable StratisError stratisError) {
            Log.d(TAG, "stratisDeviceActivationDidPostEvent");
            if (stratisError == null) {
                Log.d(TAG, deviceActivationEvent.toString());
                if (deviceActivationEvent == DeviceActivationEvent.COMPLETE) {
                    callbackSuccess(callbackContext, null);
                    responseSent = true;
                }
            } else {
                callbackError(callbackContext, stratisError.getDebugMessage());
                responseSent = true;
            }
        }
    }

    /* End command functions */

    /* Begin utility functions */

    private boolean isValidServerEnvironment(String serverEnvironmentString) {
        // Valid serverEnvironmentStrings:
        // DEV, SANDBOX, TEST, PROD
        // see https://developer.stratisiot.com/sdk/v4/android/com.stratisiot.stratissdk.constants/-server-environment/ for more info
        try {
            ServerEnvironment.valueOf(serverEnvironmentString);
            return true;
        } catch (IllegalArgumentException e) {
            return false;
        }
    }


    private JSONArray getLocksAsJson(Collection<StratisLock> locks) {
        Log.d(TAG, "getLocksAsJson");
        Bugsnag.leaveBreadcrumb("getLocksAsJson");

        JSONArray locksJsonArray = new JSONArray();

        for (StratisLock lock : locks) {
            Log.d(TAG, "serializing lock to JSON...");
            Bugsnag.leaveBreadcrumb("serializing lock to JSON...");
            JSONObject jsonObject = new JSONObject();
            try {
                jsonObject.put("isActionable", lock.isActionable());
            } catch (JSONException e) {
                e.printStackTrace();
            }
            try {
                jsonObject.put("identifier", lock.getIdentifier());
            } catch (JSONException e) {
                e.printStackTrace();
            }
            try {
                jsonObject.put("name", lock.getName());
            } catch (JSONException e) {
                e.printStackTrace();
            }
            try {
                jsonObject.put("model", lock.getModel());
            } catch (JSONException e) {
                e.printStackTrace();
            }
            locksJsonArray.put(jsonObject);
        }

        Log.d(TAG, "getLocksAsJson done");
        Bugsnag.leaveBreadcrumb("getLocksAsJson done");

        return locksJsonArray;
    }


    private void callbackSuccess(CallbackContext callbackContext, JSONArray locksJSON) {
        JSONObject r = new JSONObject();
        try {
            r.put("success", 1);
            if (locksJSON != null) {
                r.put("locks", locksJSON);
            }
        } catch (JSONException e) {
            e.printStackTrace();
        }
        callbackContext.success(r.toString());
    }


    private void callbackError(CallbackContext callbackContext, String message) {
        Log.e(TAG, message);
        bugsnagNotify(message);
        JSONObject r = new JSONObject();
        try {
            r.put("success", 0);
            if (message != null && message.length() > 0) {
                r.put("error", message);
            }
        } catch (JSONException e) {
            e.printStackTrace();
        }
        callbackContext.error(r.toString());
    }


    private void bugsnagNotify(String errorMessage) {
        class lockException extends Exception {
            public lockException(String message) {
                super(message);
            }
        }
        Bugsnag.notify(new lockException(errorMessage));
    }

    /* End utility functions*/

}

File: src/ios/OnCellStratis.swift
import StratisSDK
import Bugsnag

@objc(OnCellStratis) class OnCellStratis : CDVPlugin {
    
    var stratisSdk: StratisSDK!
    var onCellStratisDeviceAccessDelegate: OnCellStratisDeviceAccessDelegate!
    var onCellStratisDeviceDiscoveryDelegate: OnCellStratisDeviceDiscoveryDelegate!
    var accessibleLocks: [StratisLock] = []
    var discoveredLocks: [StratisLock] = []
    
    // Begin command functions
    
    @objc(initSDK:)
    func initSDK(command: CDVInvokedUrlCommand) {
        NSLog("initSDK")
        
        let serverEnvironmentString = command.arguments[0] as? String ?? ""
        let accessToken = command.arguments[1] as? String ?? ""
        let propertyID = command.arguments[2] as? String ?? ""

        Bugsnag.setBreadcrumbCapacity(50)
        Bugsnag.leaveBreadcrumb(withMessage: "Initializing Stratis SDK serverEnvironmentString: \(serverEnvironmentString), accessToken: \(accessToken), propertyId: \(propertyID)")

        var serverEnv = ServerEnvironment.SANDBOX // Sandbox by default
        if serverEnvironmentString == "PROD" {
            serverEnv = ServerEnvironment.PROD;
        }

        let configuration = Configuration(
            serverEnvironment: serverEnv,
            accessToken: accessToken,
            propertyID: propertyID,
            remoteLoggingEnabled: true,
            loggingMetadata: ["app": "OnCell"]
        )

//        if stratisSdk == nil { // workaround for SDK bug where multiple StratisSDK inits cause lock activation to stall
            stratisSdk = StratisSDK(configuration: configuration)
//        }
        
        self.accessibleLocks = [StratisLock]()
        self.discoveredLocks = [StratisLock]()

        callbackSuccess(callbackId: command.callbackId, locksJSON: nil)
    }
    
    @objc(getLocks:)
    func getLocks(command: CDVInvokedUrlCommand) {
        NSLog("getLocks")

        if self.stratisSdk != nil {
            self.accessibleLocks.removeAll()
            
            self.onCellStratisDeviceAccessDelegate = OnCellStratisDeviceAccessDelegate(onCellStratis: self, callbackId: command.callbackId)
            stratisSdk.deviceAccessDelegate = self.onCellStratisDeviceAccessDelegate

            stratisSdk.fetchAccessibleDevices()
        } else {
            callbackError(callbackId: command.callbackId, errorMessage: "Cannot get lock authorization before initializing StratisSDK")
        }
    }
    
    class OnCellStratisDeviceAccessDelegate: StratisDeviceAccessDelegate {
        var onCellStratis: OnCellStratis!
        var callbackId = ""
        
        init(onCellStratis: OnCellStratis, callbackId: String) {
            self.onCellStratis = onCellStratis
            self.callbackId = callbackId
        }
        
        func stratisDeviceAccessRequestCompleted(_ stratisSDK: StratisSDK, devices: [StratisLock], error: StratisError?) {
            NSLog("stratisDeviceAccessRequestCompleted result: \(devices)")
            if error == nil {
                onCellStratis.accessibleLocks = devices
                if onCellStratis.accessibleLocks.isEmpty {
                    onCellStratis.bugsnagNotify(exceptionName: "lockException", exceptionReason: "No accessible locks found")
                }
                let locksJson = onCellStratis.getLocksAsJson(locks: onCellStratis.accessibleLocks)
                Bugsnag.leaveBreadcrumb(withMessage: "accessibleLocks: \(locksJson)")
                onCellStratis.callbackSuccess(callbackId: callbackId, locksJSON: locksJson)
            } else {
                onCellStratis.callbackError(callbackId: callbackId, errorMessage: error!.debugDescription)
            }
        }
    }
    

    @objc(scanLocks:)
    func scanLocks(command: CDVInvokedUrlCommand) {
        NSLog("scanLocks")
        
        if !self.accessibleLocks.isEmpty {
            self.discoveredLocks.removeAll()
            
            self.onCellStratisDeviceDiscoveryDelegate = OnCellStratisDeviceDiscoveryDelegate(onCellStratis: self, callbackId: command.callbackId)
            stratisSdk.deviceDiscoveryDelegate = self.onCellStratisDeviceDiscoveryDelegate
            
            stratisSdk.discoverActionableDevices(self.accessibleLocks)
        } else {
            callbackError(callbackId: command.callbackId, errorMessage: "Cannot scan locks before getting lock authorization")
        }
    }
    
    class OnCellStratisDeviceDiscoveryDelegate: StratisDeviceDiscoveryDelegate {
        var onCellStratis: OnCellStratis!
        var callbackId = ""
        
        init(onCellStratis: OnCellStratis, callbackId: String) {
            self.onCellStratis = onCellStratis
            self.callbackId = callbackId
        }
        
        func stratisDiscoveredDevices(_ stratisSDK: StratisSDK, devices: [StratisLock]) {
            NSLog("stratisDiscoveredDevices result: \(devices)")
            onCellStratis.discoveredLocks.append(contentsOf: devices)
        }

        func stratisDiscoveryEncounteredError(_ stratisSDK: StratisSDK, error: StratisError) {
            NSLog("stratisDiscoveryEncounteredError: \(error)")
            onCellStratis.callbackError(callbackId: callbackId, errorMessage: error.debugDescription)
        }

        func stratisDiscoveryCompleted(_ stratisSDK: StratisSDK) {
            NSLog("stratisDiscoveryCompleted")
            if onCellStratis.discoveredLocks.isEmpty {
                onCellStratis.bugsnagNotify(exceptionName: "lockException", exceptionReason: "No scanned locks discovered")
            }
            let locksJson = onCellStratis.getLocksAsJson(locks: onCellStratis.discoveredLocks)
            Bugsnag.leaveBreadcrumb(withMessage: "discoveredLocks: \(locksJson)")
            onCellStratis.callbackSuccess(callbackId: callbackId, locksJSON: locksJson)
        }
    }

    @objc(activateLock:)
    func activateLock(command: CDVInvokedUrlCommand) {
        
        let lockId = command.arguments[0] as? String ?? ""
        let appointmentId = command.arguments[1] as? String ?? ""
        NSLog("activateLock lockId: \(lockId) appointmentId: \(appointmentId)")
        Bugsnag.leaveBreadcrumb(withMessage: "lockId: \(lockId)")
        Bugsnag.leaveBreadcrumb(withMessage: "appointmentId: \(appointmentId)")
        
        if !self.discoveredLocks.isEmpty {
            var foundLock = false
            for lock in self.discoveredLocks {
                if lock.identifier == lockId {
                    let deviceToActivate = lock
                    foundLock = true
                    let onCellStratisDeviceActivationDelegate = OnCellStratisDeviceActivationDelegate(onCellStratis: self, callbackId: command.callbackId)
                    deviceToActivate.activationDelegate = onCellStratisDeviceActivationDelegate
                    NSLog("Calling activate() on lock \(lockId)...")
                    deviceToActivate.activate()
                    // return error to JS and send log to Bugsnag if we haven't unlocked device in 20 seconds
                    onCellStratisDeviceActivationDelegate.startActivationTimer()
                    break
                }
            }
            if !foundLock {
                callbackError(callbackId: command.callbackId, errorMessage: "Cannot find lock to activate")
            }
        } else {
            callbackError(callbackId: command.callbackId, errorMessage: "No discovered locks to activate")
        }
    }
    
    class OnCellStratisDeviceActivationDelegate: StratisDeviceActivationDelegate {
        var onCellStratis: OnCellStratis!
        var callbackId = ""
        var responseSent = false
        
        init(onCellStratis: OnCellStratis, callbackId: String) {
            self.onCellStratis = onCellStratis
            self.callbackId = callbackId
        }
        
        func stratisDeviceActivationDidPostEvent(_ event: StratisLock.ActivationEvent, forDevice device: StratisLock, withError error: StratisError?) {
            NSLog("stratisDeviceActivationDidPostEvent with callbackId: \(self.callbackId), event: \(event), forDevice: \(device), error: \(String(describing: error))")
            if error == nil {
                if event.rawValue == ActivationEvent.activationStarted.rawValue {
                    NSLog("activationStarted");
                } else if event.rawValue == ActivationEvent.activationComplete.rawValue {
                    NSLog("activationComplete");
                    onCellStratis.callbackSuccess(callbackId: callbackId, locksJSON: nil)
                    responseSent = true;
                } else if event.rawValue == ActivationEvent.presentDeviceToLock.rawValue {
                    NSLog("presentDeviceToLock");
                }
            } else {
                onCellStratis.callbackError(callbackId: callbackId, errorMessage: error.debugDescription)
                responseSent = true;
            }
        }
        
        func startActivationTimer() {
            DispatchQueue.main.asyncAfter(deadline: .now() + 20) {
                if !self.responseSent { // If we haven't responded to the frontend by the timeout (20s)
                    self.onCellStratis.callbackError(callbackId: self.callbackId, errorMessage: "Lock activation took too long")
                }
            }
        }
    }
    
    // End command functions
    
    // Begin utility functions
    
    func callbackSuccess(callbackId: String, locksJSON: String?) {
        var message = "{\"success\": 1}"
        if locksJSON != nil {
            message = "{\"success\": 1, \"locks\": \(locksJSON!)}"
        }
        let pluginResult = CDVPluginResult(
            status: CDVCommandStatus_OK,
            messageAs: message
        )
        self.commandDelegate!.send(
            pluginResult,
            callbackId: callbackId
        )
    }
    
    func callbackError(callbackId: String, errorMessage: String) {
        NSLog(errorMessage)
        bugsnagNotify(exceptionName: "LockException", exceptionReason: errorMessage)
        let message = "{\"success\": 0, \"error\": \"\(errorMessage)\"}"
        let pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR,
            messageAs: message
        )
        self.commandDelegate!.send(pluginResult, callbackId: callbackId)
    }
    
    func bugsnagNotify(exceptionName:String, exceptionReason:String) {
        let exception = NSException(name:NSExceptionName(rawValue: exceptionName),
                                    reason:exceptionReason,
                                    userInfo:nil)
        Bugsnag.notify(exception)
    }
    
    func getLocksAsJson(locks: [StratisLock]) -> String {
        var jsonString = "[]"
        if !locks.isEmpty {
            var codableLocks = locks.compactMap({ $0.toCodable() })
            codableLocks.sort(by: {$0.rssi > $1.rssi}) // sort by RSSI in descending order
            let encoder = JSONEncoder()
            if let jsonData = try? encoder.encode(codableLocks) {
                jsonString = String(data: jsonData, encoding: .utf8) ?? "[]"
            }
        }
        return jsonString
    }
    
    // End utility functions
    
}


struct CodableLock: Codable {
    var identifier: String
    var name: String
    var model: String
    var isActionable: Bool
    var rssi: Int
}


extension StratisLock {
    
    func toCodable() -> CodableLock{
        var rssi = -100;
        // Add rssi to CodableLock if available
        if let bleLock = self as? BLELock {
            rssi = bleLock.rssi?.intValue ?? -100;
        }
        let codableLock = CodableLock(
            identifier: self.identifier,
            name: self.name,
            model: self.model,
            isActionable: self.actionable,
            rssi: rssi
        )
        
        return codableLock
    }
}
